<a href="https://colab.research.google.com/github/RanielSosa/predicciones-de-ventas/blob/main/Predicciones_de_ventas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import pandas as pd
filename = '/content/drive/MyDrive/Colab Notebooks/Docs/sales_predictions.csv'
df_sales = pd.read_csv(filename)
df_sales.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [3]:
df_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


1) ¿Cuántas filas y columnas?
  8523 filas y 12 columnas.

##2) ¿Cuáles son los tipos de datos en cada variable?
tipos: object, float64 and int64

In [ ]:
len(df_sales)

8523

##3) ¿Hay duplicados? Si es el caso, eliminen algunos duplicados.

No existen duplicados

In [ ]:
df_sales.duplicated().sum()

0

##Ejercicio 4

In [57]:
nulos_Weight=df_sales['Item_Weight'].isnull()
nulos_Size=df_sales['Outlet_Size'].isnull()
df_sales.isnull().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

4) Identifiquen los valores faltantes.
  Item_Weight y Outlet_Size son las unicas columnas con valores faltantes.

##Punto 5

In [53]:
df_sales.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [54]:
nulos_Weight

0       False
1       False
2       False
3       False
4       False
        ...  
8518    False
8519    False
8520    False
8521    False
8522    False
Name: Item_Weight, Length: 8523, dtype: bool

### ITEM WEIGHT

estoy intentando encontrar el patron de que relacione los articulos con falta de informacion con los articulos que si tienen toda la informacion. hasta ahora, ya se que a los que les falta informacion de una columna no le falta de la otra columna.

In [55]:
Item_Weight_NA=df_sales.loc[nulos_Weight,:]
Item_Weight_NA.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                     0
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [91]:
Item_Weight_NA['Item_Type'].value_counts()

Fruits and Vegetables    213
Snack Foods              212
Household                151
Frozen Foods             138
Dairy                    116
Baking Goods             112
Canned                   110
Health and Hygiene        90
Meat                      88
Soft Drinks               71
Breads                    47
Others                    32
Hard Drinks               31
Breakfast                 21
Starchy Foods             18
Seafood                   13
Name: Item_Type, dtype: int64

In [87]:
Item_Weight_NA.nunique()

Item_Identifier              1142
Item_Weight                     0
Item_Fat_Content                2
Item_Visibility              1362
Item_Type                      16
Item_MRP                     1371
Outlet_Identifier               2
Outlet_Establishment_Year       1
Outlet_Size                     2
Outlet_Location_Type            2
Outlet_Type                     2
Item_Outlet_Sales            1183
dtype: int64

In [ ]:
Item_Weight_NA['Item_Fat_Content']=Item_Weight_NA['Item_Fat_Content'].replace('LF', 'Low Fat')
Item_Weight_NA['Item_Fat_Content']=Item_Weight_NA['Item_Fat_Content'].replace('low fat', 'Low Fat')
Item_Weight_NA['Item_Fat_Content']=Item_Weight_NA['Item_Fat_Content'].replace('reg', 'Regular')

In [90]:
#Item_Weight_NA.groupby(['Outlet_Type','Outlet_Location_Type','Item_Type']).count()
Item_Weight_NA.groupby(['Outlet_Type','Outlet_Location_Type','Item_Fat_Content','Outlet_Identifier','Outlet_Size']).count()

Item_Identifier  \
Outlet_Type       Outlet_Location_Type Item_Fat_Content Outlet_Identifier Outlet_Size                    
Grocery Store     Tier 1               Low Fat          OUT019            Small                    346   
                                       Regular          OUT019            Small                    182   
Supermarket Type3 Tier 3               Low Fat          OUT027            Medium                   605   
                                       Regular          OUT027            Medium                   330   

                                                                                       Item_Weight  \
Outlet_Type       Outlet_Location_Type Item_Fat_Content Outlet_Identifier Outlet_Size                
Grocery Store     Tier 1               Low Fat          OUT019            Small                  0   
                                       Regular          OUT019            Small                  0   
Supermarket Type3 Tier 3               Low Fat          OUT027            Medium                 0   
                                       Regular          OUT027            Medium                 0   

                                                                                       Item_Visibility  \
Outlet_Type       Outlet_Location_Type Item_Fat_Content Outlet_Identifier Outlet_Size                    
Grocery Store     Tier 1               Low Fat          OUT019            Small                    346   
                                       Regular          OUT019            Small                    182   
Supermarket Type3 Tier 3               Low Fat          OUT027            Medium                   605   
                                       Regular          OUT027            Medium                   330   

                                                                                       Item_Type  \
Outlet_Type       Outlet_Location_Type Item_Fat_Content Outlet_Identifier Outlet_Size              
Grocery Store     Tier 1               Low Fat          OUT019            Small              346   
                                       Regular          OUT019            Small              182   
Supermarket Type3 Tier 3               Low Fat          OUT027            Medium             605   
                                       Regular          OUT027            Medium             330   

                                                                                       Item_MRP  \
Outlet_Type       Outlet_Location_Type Item_Fat_Content Outlet_Identifier Outlet_Size             
Grocery Store     Tier 1               Low Fat          OUT019            Small             346   
                                       Regular          OUT019            Small             182   
Supermarket Type3 Tier 3               Low Fat          OUT027            Medium            605   
                                       Regular          OUT027            Medium            330   

                                                                                       Outlet_Establishment_Year  \
Outlet_Type       Outlet_Location_Type Item_Fat_Content Outlet_Identifier Outlet_Size                              
Grocery Store     Tier 1               Low Fat          OUT019            Small                              346   
                                       Regular          OUT019            Small                              182   
Supermarket Type3 Tier 3               Low Fat          OUT027            Medium                             605   
                                       Regular          OUT027            Medium                             330   

                                                                                       Item_Outlet_Sales  
Outlet_Type       Outlet_Location_Type Item_Fat_Content Outlet_Identifier Outlet_Size                     
Grocery Store     Tier 1               Low Fat          OUT019            Small              

### OUTLET SIZE

In [58]:
Outlet_Size_NA=df_sales.loc[nulos_Size,:]
Outlet_Size_NA.isna().sum()

Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

### Continuacion

In [4]:
df_sales.nunique()

Item_Identifier              1559
Item_Weight                   415
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

In [8]:
df_sales[['Item_Weight','Outlet_Size']]

Item_Weight Outlet_Size
0           9.300      Medium
1           5.920      Medium
2          17.500      Medium
3          19.200         NaN
4           8.930        High
...           ...         ...
8518        6.865        High
8519        8.380         NaN
8520       10.600       Small
8521        7.210      Medium
8522       14.800       Small

[8523 rows x 2 columns]

In [9]:
valores=df_sales.groupby(['Item_Type','Outlet_Type'])[['Item_Identifier','Item_Weight','Outlet_Size']].count()
valores

Item_Identifier  Item_Weight  Outlet_Size
Item_Type     Outlet_Type                                                 
Baking Goods  Grocery Store                   85           42           43
              Supermarket Type1              426          426          283
              Supermarket Type2               68           68           68
              Supermarket Type3               69            0           69
Breads        Grocery Store                   33           17           16
...                                          ...          ...          ...
Soft Drinks   Supermarket Type3               45            0           45
Starchy Foods Grocery Store                   13            9            4
              Supermarket Type1              104          104           70
              Supermarket Type2               17           17           17
              Supermarket Type3               14            0           14

[64 rows x 3 columns]

5) Decidan cómo abordar los valores faltantes y cómo hacerlo. (Esto requiere sus criterios, así que expliquen su elección).


Mi propuesta es agrupar los registros por tipo de producto y otras variables que me ayuden a identificar un patron en el producto no identificado y una vez encontrado el patron realizar la imputacion de los datos usando como referencia los valores similares a los articulos con valores faltantes, pero aun no he podido encontrar la forma correcta de relacionar los articulos en grupos que me puedan dar informacion muy especifica del conjunto de datos flatantes.

estoy usando el metodo 'Nunique' para identificar cuales columnas tienen datos utilizables pero con pocos valores unicos.

Tambien estoy agrupando por el set de datos con menos valores unicos para entender que informacion obtengo y luego tomar varios registros al azar de pueba para validar que la relacion es la mas eficiente para imputar y en caso de que no exista nada que relacione los registros con datos faltantes con los registros con todos sus datos entonces esos datos hay que eliminarlos.

##6) Confirmen que no hay valores faltantes después de abordarlos.

Si el planteamiento es correcto, no existen datos faltantes, aunque pueda que existan unos cuantos registros menos que la cantidad inicial

7) Encuentren y arreglen alguna categoría inconsistente de datos (example: fix cat, Cat, and cats so that they are consistent) 

In [ ]:
df_sales.nunique()

Item_Identifier              1559
Item_Weight                   415
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

para determinar cual categoria contenia valores mal escritos, volvi a ver las columnas con valores unicos y luego fuí probando con las que tenian valores unicos por debajo de 10

In [83]:
df_sales['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

In [82]:
df_sales['Item_Fat_Content']=df_sales['Item_Fat_Content'].replace('LF', 'Low Fat')
df_sales['Item_Fat_Content']=df_sales['Item_Fat_Content'].replace('low fat', 'Low Fat')
df_sales['Item_Fat_Content']=df_sales['Item_Fat_Content'].replace('reg', 'Regular')

In [ ]:
df_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


8) Para cualquier columna numérica, obtengan las estadísticas resumidas para cada uno (mínimo, máximo y media)


In [ ]:
df_sales['Outlet_Establishment_Year'].mean()

1997.8318667135984

In [ ]:
df_sales['Outlet_Establishment_Year'].min()

1985

In [ ]:
df_sales['Outlet_Establishment_Year'].max()

2009